[exporter/datasetexporter] Various improvements and fixes for the DataSet Log Exporter plugin

tomaz-s1 opened the pull request:
# Related issue
#20660

## Remove OtelExporter - Log - prefix from all the DataSet event message fields

This prefix was added in the past for testing purposes, but it's not desired.

## Correctly handle LogRecord severity

The code has been updated to correctly map ``LogRecord.SeverityNumber`` and ``LogRecord.SeverityText`` field value to the DataSet event log event severity (https://app.scalyr.com/help/parsing-logs#specialAttrs, https://github.com/scalyr/logstash-output-scalyr/blob/master/lib/logstash/outputs/scalyr.rb#L70)

Redundant `severity.number` and `severity.text` attribute has also been removed from DataSet event.

## Remove redundant flags and flags.is_sampled attribute

I removed redundant `flags` and `flags.is_sampled` attribute from the DataSet events. DataSet doesn't do anything with those fields and they are also redundant - if event is ingested into DataSet and displayed in the UI this means it has already been sampled so `flags.is_sampled` tells us no new information.

zdaratom-s1 committed:
DSET-3998 - export Logs resource info based on     export_resource_info_on_event configuration

zdaratom-s1 committed:
DSET-3998 - simplify

zdaratom-s1 committed:
DSET-3998 - improve docs

tomaz-s1 committed:
Fix log exporter to set AddEvents Event timestamp (ts) field to event
observed timetamp in case LogRecord doesn't contain timestamp.

Even though ObservedTimestamp should always be present, we fall back to
current time in case it's not.

In addition to that, remove duplicate and redundant "timestamp"
attribute from the AddEvents event.

zdaratom-s1 committed:
Merge pull request #1 from zdaratom/DSET-3998

DSET-3998 feat: export Logs resource info based on export_resource_info_on_event configuration

tomaz-s1 committed:
Merge branch 'datasetexporter-latest' of github.com:scalyr/opentelemetry-collector-contrib into event_no_timestamp_fix_use_occurence_time

tomaz-s1 committed:
Add additional assertions.

tomaz-s1 committed:
Remove dummy debug logs.

tomaz-s1 committed:
Merge pull request #2 from scalyr/event_no_timestamp_fix_use_occurence_time

[DSET-4071] [DSET-4007] Fix timestamp handling for LogRecord objects without a timestamp

zdaratom-s1 committed:
Create export-logs-resource-info-based-configuration

zdaratom-s1 committed:
address PR notes - fix changelog gen

zdaratom-s1 committed:
fix docs typo

zdaratom-s1 committed:
fix changelog file suffix

tomaz-s1 committed:
Remove "OtelExporter - Log -" suffix from the event "message" field.
This prefix is nor desired nor wanted.

tomaz-s1 committed:
Implement severity field handling and make sure we currently map OTel
severity to DataSet severity.

Also remove "severity.text" and "severity.number" field from the event
since it's redundant - we already have event severity (sev) field value.

tomaz-s1 committed:
Remove flags and flags.is_sampled field since it provides no additional
information which is not already available (events which are sampled
will already be ingested and visible in the ui so flags.is_sampled is
redundant).

zdaratom-s1 committed:
Merge branch 'open-telemetry:main' into datasetexporter-latest

zdaratom-s1 committed:
Merge branch 'open-telemetry:main' into datasetexporter-latest

tomaz-s1 committed:
Update metadata and indicate plugin is distributed as part of otel
contrib.

tomaz-s1 committed:
Move default severity to be a module level constant instead of function
level constant.

tomaz-s1 committed:
Refactor / split otelSeverityToDataSetSeverity to two functions.

tomaz-s1 committed:
Also add test cases for invalid values.

tomaz-s1 committed:
Make it a module level constant.

tomaz-s1 committed:
Make dataset log levels constants for easier readability and maintaince
and update affected code and tests.

tomaz-s1 committed:
Re-generate readme.

tomaz-s1 committed:
Per PR review, add "map" suffix to the function names.

tomaz-s1 committed:
For readability, split different test scenarios in different test
functions. Also move functionality for making a test record into a
utility function.

tomaz-s1 committed:
Add changelog entry file.

tomaz-s1 committed:
Merge pull request #3 from scalyr/dataset_log_exporter_improvements

[DSET-4037] [DSET-4014] [DSET-4034] [DSET-4034] [DSET-4098] Dataset log exporter improvements

tomaz-s1 committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into dataset_log_exporter_improvements_fixes

tomaz-s1 committed:
Merge branch 'datasetexporter-latest' into dataset_log_exporter_improvements_fixes

tomaz-s1 commented on .chloggen/export-logs-resource-info-based-configuration.yaml:
```diff
@@ -0,0 +1,20 @@
+component: datasetexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Allow include Logs resource info export to DataSet based on new export_resource_info_on_event configuration. Fix timestamp handling."
```
This is a left over from another PR from @zdaratom which for some reason was missing from his PR.

atoulme replied:
not sure I follow, what is this changelog entry doing here?

tomaz-s1 replied:
We missed / forgot to add changelog as part of that PR (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23250), so I added it here.

atoulme replied:
Please create a chore pr to add the changelog asap

atoulme replied:
I see the changelog entry there: .chloggen/export-logs-resource-info-based-configuration.yaml so I don’t think it was missed.

tomaz-s1 replied:
@atoulme I'm not sure I understand your comment. I don't see that changelog file in the main branch (https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/.chloggen), I only see it in this one where I added it.

Am I missing something? Thanks.

tomaz-s1 replied:
I went ahead and removed that changelo file.

tomaz-s1 commented:
Not 100% sure, but the failure here (https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/5375815914/jobs/9752150013?pr=23672) also doesn't seem to be related to my changes.

I would re-run it to see if it's a flakey tests, but I don't have permission to do that.

dmitryax commented:
@tomaz-s1 the failing integration test was fixed in the main branch. Please rebase

tomaz-s1 committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into dataset_log_exporter_improvements_fixes

tomaz-s1 commented:
@dmitryax Thanks. Done. All the checks appear to be passing now.

tomaz-s1 commented:
@dmitryax When you get a chance, please let me know if there is anything else which needs to be done on my side. Thanks a lot!

tomaz-s1 commented:
@atoulme As a code owner, could you please review when you get a chance? Thanks.

atoulme commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +81,95 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
+	}
+
+	// TODO: We should log in case we see invalid severity, but right now, afaik, we / OTEL
+	// don't have a concept of "rate limited" logging. We don't want to log every single
```
Otel offers probabilistic head sampling for logs and traces, you can use that. There is no need for a TODO entry here as this is best captured with an issue on the github repository as a follow up instead.

tomaz-s1 replied:
Per my reading and understanding of this sampling processor (https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/probabilisticsamplerprocessor), the processor is not flexible enough to only sample specific log messages.

So I could define a processor rule to sample all the log exporter logging messages, but not just this specific log message.

Other log messages are not that noisy / verbose so I would only want to sample this specific log message.

And I don't really want to use debug log level here either. I still want to log it, ideally, under warning log level, but only emit one message every X seconds in case there are lot of messages like that to prevent noise, overhead, etc.

Please correct me if I'm wrong / misunderstanding it.

Or is there some other existing code pattern / API in the code base I could use to do that?

tomaz-s1 replied:
If not, to unblock the PR, I'm also fine with removing this TODO comment for now and open an issue instead.

atoulme replied:
Please do.

atoulme commented:
OK, fwiw it's much easier if you run 4 PRs instead of one. Please do peel off at least the metadata changes as a separate `[chore]` PR that we can easily merge. I can't judge the changes much except for the TODO entry, see my comment. Please address my comment and I will approve the PR.

tomaz-s1 commented:
@atoulme Thanks for a review.

And yeah, I agree - it would make it a bit easier to review if those were 4 separate PRs, but per PR description above, I tried first with combining all 4 changes together first since they are small, related and makes it a bit easier for us. 

But in case it's needed, I can definitely split it out :)

Having said that, I will split out metadata change in a separate PR.

tomaz-s1 committed:
Revert metadata change since it will be moved in a separate standalone
PR.

tomaz-s1 committed:
Update changelog entry, make sure we reference related PR.

tomaz-s1 commented:
Moved metadata change in a new and separate PR - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23885.

tomaz-s1 committed:
Remove TODO annotations / comments from code.

Per PR review, we want to create issues for those instead (once the PR
is merged).

tomaz-s1 commented:
Integration tests failure doesn't seem to be related to my changes. Will sync this branch again with latest main which will trigger another build, to see if it helps.

tomaz-s1 committed:
Merge branch 'main' of https://github.com/open-telemetry/opentelemetry-collector-contrib into dataset_log_exporter_improvements_fixes

tomaz-s1 commented:
@atoulme I removed the TODO comments / annotations and synced up this branch with latest main. It looks like everything is passing now. Can you please check again when you get a chance? Thanks.

tomaz-s1 committed:
Reduce the scope and simplify the PR and remove the changelog entry for
a different PR where we forgot to add it.

This changelog will be added as part of a separate PR.

atoulme reviewed (approved):
LGTM
